File: backend/app/api/v1/endpoints/budgets.py
import uuid

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.base import get_db
from app.models.user import User
from app.services.budget import BudgetService
from app.schemas.budget import BudgetCreate, BudgetUpdate, BudgetResponse

router = APIRouter(prefix="/budgets", tags=["Budgets"])


@router.get("/", response_model=list[BudgetResponse])
async def list_budgets(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await BudgetService(db).list_for_user(current_user.id)


@router.post("/", response_model=BudgetResponse, status_code=201)
async def create_budget(
    data: BudgetCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await BudgetService(db).create(data.model_dump(), current_user.id)


@router.get("/{budget_id}", response_model=BudgetResponse)
async def get_budget(
    budget_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await BudgetService(db).get(budget_id, current_user.id)


@router.patch("/{budget_id}", response_model=BudgetResponse)
async def update_budget(
    budget_id: uuid.UUID,
    data: BudgetUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = BudgetService(db)
    budget = await svc.get(budget_id, current_user.id)
    return await svc.update(budget, data.model_dump(exclude_unset=True))


@router.delete("/{budget_id}", status_code=204)
async def delete_budget(
    budget_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    svc = BudgetService(db)
    budget = await svc.get(budget_id, current_user.id)
    await svc.delete(budget)

File: frontend/src/lib/api/budgets.ts
import { api } from "@/api/client";
import type { Budget } from "@/types";

export const budgetsApi = {
  list: async (): Promise<Budget[]> => {
    const res = await api.get("/v1/budgets/");
    return res.data;
  },

  get: async (id: string): Promise<Budget> => {
    const res = await api.get(`/v1/budgets/${id}`);
    return res.data;
  },

  create: async (data: Record<string, unknown>): Promise<Budget> => {
    const res = await api.post("/v1/budgets/", data);
    return res.data;
  },

  update: async (id: string, data: Record<string, unknown>): Promise<Budget> => {
    const res = await api.patch(`/v1/budgets/${id}`, data);
    return res.data;
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/v1/budgets/${id}`);
  },
};
